toolchain/check/impl.cpp

```diff
@@ -134,27 +134,51 @@ static auto CheckAssociatedFunctionImplementation(
   return impl_decl_id;
 }
 
-// Builds a witness that the specified impl implements the given interface.
-static auto BuildInterfaceWitness(
-    Context& context, const SemIR::Impl& impl, SemIR::TypeId facet_type_id,
-    SemIR::FacetTypeInfo::ImplsConstraint interface_type,
-    llvm::SmallVectorImpl<SemIR::InstId>& used_decl_ids) -> SemIR::InstId {
-  const auto& interface = context.interfaces().Get(interface_type.interface_id);
+auto ImplWitnessForDeclaration(Context& context, const SemIR::Impl& impl)
+    -> SemIR::InstId {
+  auto facet_type_id = context.GetTypeIdForTypeInst(impl.constraint_id);
+  if (facet_type_id == SemIR::ErrorInst::SingletonTypeId) {
+    return SemIR::ErrorInst::SingletonInstId;
+  }
+  auto facet_type = context.types().TryGetAs<SemIR::FacetType>(facet_type_id);
+  if (!facet_type) {
+    CARBON_DIAGNOSTIC(ImplAsNonFacetType, Error, "impl as non-facet-type");
```

Comment:
```suggestion
    CARBON_DIAGNOSTIC(ImplAsNonFacetType, Error, "impl as non-facet type");
```
Do you think it'd be useful to include the type `facet_type_id` in this diagnostic?

Reply:
Used `impl.constraint_id` instead of `facet_type_id` so we can eventually use the source's spelling of that type.